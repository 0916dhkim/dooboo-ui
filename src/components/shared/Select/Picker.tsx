import * as Select from './index';

import { FlatList, Modal, NativeScrollEvent, NativeSyntheticEvent, ViewStyle } from 'react-native';
import React, { useCallback, useRef } from 'react';

import styled from 'styled-components/native';

export const ERR_MSG_ITEMS_REQUIRED = `items is required. 
Please put items of the form [{value: string, label?: string}, ...].`;

export enum PICKER_TEST_ID {
  MODAL = 'modal',
  CLOSE = 'close',
  LIST_WRAPPER = 'list-wrapper',
  LIST = 'list',
  ITEM = 'item'
}

/**
 * interfaces
 */
interface Item {
  label?: string;
  value: string;
}

interface Props extends Partial<Select.Props> {
  listOpen: Select.ListOpen;
  setListOpen: (listOpen: Select.ListOpen) => void;
}

type ItemLayout = {
  length: number;
  offset: number;
  index: number;
}

/**
 * constants
 */
const ITEM_HEIGHT = 40;
const ITEM_LIST_HEIGHT = 240;

/**
 * styled components
 */
const CloseTouchableOpacity = styled.TouchableOpacity`
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: #efefef55;
`;

const ItemListContainer = styled.View`
  height: 240px; 
  width: 200px;
  border-radius: 1px;
`;

const ItemFlatList = styled(FlatList as new () => FlatList<Item>)`
  width: 100%;
  height: 100%;
`;

const ItemView = styled.TouchableOpacity`
  width: 100%;
  height: 40px;
  justify-content: center;
  align-items: center;
`;

const ItemText = styled.Text`
  font-size: 14px;
`;

const EmptyView = styled.View`
  height: 80px;
  width: 100%;
`;

const Line = styled.View`
  position: absolute;
  height: 1px;
  width: 100%;
  background-color: grey;
`;

const getListStyle = ({ theme, itemListStyle, listOpen, itemHeight }): ViewStyle => {
  const listStyle: ViewStyle = {
    position: 'absolute',
    top: listOpen.y - itemHeight * 2.5,
    left: listOpen.x,
    width: listOpen.width,
    backgroundColor: Select.COLOR.WHITE,
  };

  if (itemListStyle) {
    if (itemListStyle.bottom !== undefined) delete listStyle.top;
    if (itemListStyle.right !== undefined) delete listStyle.left;
    return { ...listStyle, ...itemListStyle };
  } else {
    let themeStyle;
    switch (theme) {
      case Select.ThemeEnum.none: {
        themeStyle = {
          shadowColor: Select.COLOR.DODGERBLUE,
          shadowOffset: { width: 2, height: 3 },
          shadowOpacity: 0.2,
          shadowRadius: 5,
          elevation: 6,
        };
        break;
      }
      case Select.ThemeEnum.box: {
        themeStyle = {
          borderColor: Select.COLOR.GRAY59,
          borderWidth: 2,
        };
        break;
      }
      default: {
        themeStyle = {
          shadowColor: Select.COLOR.GRAY59,
          shadowOffset: { width: 2, height: 3 },
          shadowOpacity: 0.2,
          shadowRadius: 5,
          elevation: 6,
        };
        break;
      }
    }
    return { ...listStyle, ...themeStyle };
  }
};

function Picker({
  listOpen,
  onShow,
  onDismiss,
  setListOpen,
  onValueChange,
  selectedValue,
  items,
  itemListStyle,
  itemViewStyle,
  itemTextStyle,
  showsVerticalScrollIndicator,
  testID,
  theme,
}: Props): React.ReactElement {
  // check valid props;
  if (!items || !items.length) throw new Error(ERR_MSG_ITEMS_REQUIRED);
  const flatList = useRef<FlatList<Item>>(null);

  // constants
  const itemHeight = itemViewStyle ? itemViewStyle.height || ITEM_HEIGHT : ITEM_HEIGHT;
  const itemListHeight = itemListStyle ? itemListStyle.height || ITEM_LIST_HEIGHT : ITEM_LIST_HEIGHT;
  const itemThreshold = itemHeight / 2;
  let position = items.findIndex((item) => item.value === selectedValue);
  if (position === -1) position = 0;
  let newPosition = position;
  let isScrolling = false;

  const emptyViewHeight = (itemListHeight - itemHeight) / 2;

  const close = (): void => {
    setListOpen({
      ...listOpen,
      isOpen: false,
    });
  };

  const scrollToOffset = ({ offset }): void => {
    if (flatList && flatList.current) {
      flatList.current.scrollToOffset({ offset });
    }
  };

  const scrollToPosition = (position): void => {
    const offset = position * itemHeight;
    scrollToOffset({ offset });
  };

  const renderItem = ({ item: { value, label }, index }: {item: Item; index: number}): React.ReactElement => {
    const onPress = (): void => {
      if (value === selectedValue) close();
      else scrollToPosition(index);
    };
    return (
      <ItemView
        testID={`${testID}-${PICKER_TEST_ID.ITEM}-${index}`}
        style={itemViewStyle}
        onPress={onPress}>
        <ItemText style={itemTextStyle}>{label || value}</ItemText>
      </ItemView>
    );
  };

  const onScroll = useCallback((e: NativeSyntheticEvent<NativeScrollEvent>): void => {
    const { nativeEvent: { contentOffset: { y } } } = e;
    newPosition = Math.floor(y / itemHeight);
    if (y % itemHeight > itemThreshold) newPosition++;
    if (newPosition < 0) newPosition = 0;
    else if (newPosition >= items.length) newPosition = items.length - 1;
    if (position !== newPosition) {
      position = newPosition;
      if (onValueChange) onValueChange(items[position], position);
    }
  }, []);

  const onScrollBeginDrag = useCallback((): void => {
    isScrolling = true;
  }, []);

  const onScrollEndDrag = useCallback((): void => {
    isScrolling = false;
    setTimeout(() => {
      if (!isScrolling) scrollToPosition(position);
    }, 0);
  }, []);

  const onMomentumScrollBegin = useCallback((): void => {
    isScrolling = true;
  }, []);

  const onMomentumScrollEnd = useCallback((): void => {
    if (!isScrolling) return;
    isScrolling = false;
    setTimeout(() => {
      if (!isScrolling) scrollToPosition(position);
    }, 0);
  }, []);

  const getItemLayout = (data: Item[] | null, index: number): ItemLayout => ({
    length: 0,
    offset: itemHeight * index,
    index,
  });

  const listStyle = getListStyle({ theme, itemListStyle, listOpen, itemHeight });

  return (
    <Modal
      visible={listOpen.isOpen}
      onShow={onShow}
      onDismiss={onDismiss}
      transparent
    >
      <CloseTouchableOpacity
        testID={`${testID}-${PICKER_TEST_ID.CLOSE}`}
        onPress={close}
        activeOpacity={1}/>
      <ItemListContainer
        testID={`${testID}-${PICKER_TEST_ID.LIST_WRAPPER}`}
        style={listStyle}>
        <ItemFlatList
          ref={flatList}
          testID={`${testID}-${PICKER_TEST_ID.LIST}`}
          showsVerticalScrollIndicator={showsVerticalScrollIndicator}
          ListHeaderComponent={<EmptyView style={{ height: emptyViewHeight }}/>}
          ListFooterComponent={<EmptyView style={{ height: emptyViewHeight }}/>}
          keyExtractor={({ value }: Item): string => value}
          data={items}
          renderItem={renderItem}
          onScroll={onScroll}
          onScrollBeginDrag={onScrollBeginDrag}
          onScrollEndDrag={onScrollEndDrag}
          onMomentumScrollBegin={onMomentumScrollBegin}
          onMomentumScrollEnd={onMomentumScrollEnd}
          initialScrollIndex={position}
          getItemLayout={getItemLayout}
        />
        <Line style={{ top: emptyViewHeight }}/>
        <Line style={{ bottom: emptyViewHeight }}/>
      </ItemListContainer>
    </Modal>
  );
}

export default Picker;
