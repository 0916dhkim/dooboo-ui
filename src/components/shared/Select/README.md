# Select

- [![Npm Version](http://img.shields.io/npm/v/@dooboo-ui/native-select.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-select)
- [![Downloads](http://img.shields.io/npm/dm/@dooboo-ui/native-select.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-select)

![Select Component themes](https://user-images.githubusercontent.com/33364619/70374774-fd328c00-1939-11ea-9af4-776c6885bd41.png)

## Types

| name | describe                         | required |
| ---- | -------------------------------- | -------- |
| Item | { value: string, label: string } | value    |

## Props

|                                 | necessary | types                                | default    |
| ------------------------------- | --------- | ------------------------------------ | ---------- |
| testID                          |           | string                               |            |
| items                           | âœ“         | Item[ ]                              |            |
| style                           |           | ViewStyle                            |            |
| [theme](#theme)                 |           | enum('none', 'underbar', 'box')      | 'none'     |
| title                           |           | string                               |            |
| titleStyle                      |           | TextStyle                            |            |
| selectedValue                   |           | string                               |            |
| placeholder                     |           | string                               | ''         |
| activeOpacity                   |           | number(0 - 1.0)                      | 0.8        |
| rootViewStyle                   |           | ViewStyle                            |            |
| rootTextStyle                   |           | TextStyle                            |            |
| mode                            |           | enum('dropdown', 'dialog', 'picker') | 'dropdown' |
| itemListStyle                   |           | ViewStyle                            |            |
| itemViewStyle                   |           | ViewStyle                            |            |
| itemTextStyle                   |           | TextStyle                            |            |
| showsVerticalScrollIndicator    |           | boolean                              |            |
| disabled                        |           | boolean                              | `false`    |
| [onSelect](#onSelect)           |           | (item: Item) => void                 |            |
| [onShow](#onShow)               |           | ( ) => void                          |            |
| [onDismiss](#onDismiss)         |           | ( ) => void                          |            |
| [onValueChange](#onValueChange) |           | (item: Item, index: number) => void  |            |

### Description of `Props`

#### theme

- injecting some value to prop `theme`, the other prop `rootViewStyle` doesn't work.
  - `theme` & `rootTextStyle` props work
- if you want to `rootViewStyle` work, make `theme` prop empty

#### title

- This text data will appear on the top of select box as title or label.

#### titleTextStyle

- This text style will be applied to Title section.

#### onSelect

...

#### onShow

The **onShow** prop allows passing a function that will be called once the item list has been shown.

| type     | required |
| -------- | -------- |
| function | No       |

#### onDismiss

The **onDismiss** prop allows passing a function that will be called once the item list has been dismissed.

| type     | required |
| -------- | -------- |
| function | No       |

#### onValueChange

(mode **picker** only)
Callback for when an item is changes in picker mode. This is called with the following parameters:

item: the value prop of the item that was selected,
index: the index of the selected item in this picker

| type     | required |
| -------- | -------- |
| function | No       |

> > > > > > > add picker component and test code for picker mode

## Installation

```sh
yarn add @dooboo-ui/native
```

or

```sh
yarn add @dooboo-ui/native-select
```

## Getting started

- Import

  ```javascript
  import { Select } from '@dooboo-ui/native';
  ```

- Usage

  ```typescript
  const ITEMS = [
    { value: 'Category1', text: 'Category1' },
    { value: 'Category2', text: 'Category2' },
    { value: 'Category3', text: 'Category3' },
    { value: 'Category4', text: 'Category4' },
    { value: 'Category5', text: 'Category5' },
  ];

  const [selectedItem, setSelectedItem] = useState<Item>(null);
  const onSelect = useCallback(
    (item: Item) => {
      setSelectedItem(item);
    },
    [selectedItem],
  );
  ```

  ```jsx
  <Container>
    <Select
      theme={'underbar'}
      rootViewStyle={{
        borderBottomColor: 'black',
        borderBottomWidth: 2,
      }}
      rootTextStyle={{
        color: 'orange',
      }}
      itemStyle={{
        list: {},
        defaultItem: {
          color: 'grey',
        },
        selectedItem: {
          color: 'black',
        },
      }}
      placeholder={'select'}
      onClick={action('Clicked')}
      items={ITEMS}
      onSelect={onSelect}
      selectedItem={selectedItem}
    />
  </Container>
  ```
