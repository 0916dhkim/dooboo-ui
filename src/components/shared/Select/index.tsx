import {
  FlatList,
  Image,
  ListRenderItemInfo,
  Modal,
  StyleProp,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';
import { IC_ARR_DOWN, IC_ARR_UP } from '../Icons';
import React, { ReactElement, useCallback, useRef, useState } from 'react';
import styled, { DefaultTheme, css } from 'styled-components/native';

import { FlattenSimpleInterpolation } from 'styled-components';
import Picker from './Picker';

export enum ThemeEnum {
  disabled = 'disabled',
  blank = 'blank',
  none = 'none',
  box = 'box',
  underbar = 'underbar',
}

enum CompEnum {
  rootbox = 'rootbox',
  text = 'text',
  item = 'item',
}
enum StylePropEnum {
  bc = 'backgroundColor',
  fc = 'fontColor',
  bs = 'boxShadow',
  border = 'border',
}

export enum Mode {
  dropdown = 'dropdown',
  dialog = 'dialog',
  picker = 'picker',
}

type ThemeStyle<K extends string, T> = {
  [P in K]: T;
};

interface ThemeType {
  theme: ThemeEnum;
}

export interface Item {
  value: string;
  label?: string;
}

interface BorderStyle extends ViewStyle {
  borderColor?: string;
  borderWidth?: number;
  borderBottomColor?: string;
  borderBottomWidth?: number;
  borderLeftColor?: string;
  borderLeftWidth?: number;
  borderRightColor?: string;
  borderRightWidth?: number;
  borderTopColor?: string;
  borderTopWidth?: number;
}

interface ItemListStyle extends ViewStyle {
  height?: number;
}
interface ItemViewStyle extends ViewStyle {
  height?: number;
}

interface RootBoxTheme extends DefaultTheme {
  rootbox: {
    backgroundColor: string;
    boxShadow?: FlattenSimpleInterpolation;
    border?: BorderStyle;
  };
}

interface TextTheme extends DefaultTheme {
  text: {
    fontColor: string;
  };
}

interface ThemeType {
  theme: ThemeEnum;
}
interface Selected {
  selected: boolean;
}

export interface ListOpen {
  isOpen: boolean;
  x: number;
  y: number;
  width: number;
  height: number;
}

export const TESTID = {
  TITLETEXT: 'title-text',
  ROOTBUTTON: 'root-button',
  ROOTSELECT: 'root-select',
  ROOTTEXT: 'root-text',
  ROOTARROW: 'root-arrow',
  SELECTLISTVIEW: 'select-list-view',
  LISTITEM: 'list-item',
  MODALCLOSEVIEW: 'modal-close-view',
  PICKER: 'picker',
};

export const COLOR: {
  [key: string]: string;
} = {
  WHITE: '#ffffff',
  DODGERBLUE: '#5364ff',
  VERYLIGHTGRAY: '#cccccc',
  LIGHTGRAY: '#c8c8c8',
  BLUE: '#0000ff',
  STRONGBLUE: '#069ccd',
  GRAY3: '#080808',
  GRAY7: '#121212',
  GRAY59: '#969696',
  DARK: '#09071d',
  LIGHTBLUE: '#bcdbfb',
  BLACK: '#000000',
};

const bsCss = css`
  elevation: 1;
  shadow-color: ${COLOR.DODGERBLUE};
  shadow-offset: {
    width: 2;
    height: 2;
  }
  shadow-opacity: 0.5;
  shadow-radius: 3;
`;

export const themeStylePropCollection: ThemeStyle<
  ThemeEnum,
  RootBoxTheme | TextTheme
> = {
  disabled: {
    rootbox: {
      backgroundColor: 'transparent',
      border: {
        borderBottomColor: COLOR.LIGHTGRAY,
        borderBottomWidth: 2,
      },
    },
    text: {
      fontColor: COLOR.LIGHTGRAY,
    },
  },
  blank: {
    rootbox: {
      backgroundColor: 'transparent',
    },
    text: {
      fontColor: COLOR.DARK,
    },
  },
  none: {
    rootbox: {
      backgroundColor: COLOR.WHITE,
      boxShadow: bsCss,
    },
    text: {
      fontColor: COLOR.DARK,
    },
  },
  box: {
    rootbox: {
      backgroundColor: COLOR.WHITE,
      border: {
        borderColor: COLOR.GRAY59,
        borderWidth: 2,
      },
    },
    text: {
      fontColor: COLOR.DARK,
    },
  },
  underbar: {
    rootbox: {
      backgroundColor: COLOR.WHITE,
      border: {
        borderBottomColor: COLOR.GRAY59,
        borderBottomWidth: 2,
      },
    },
    text: {
      fontColor: COLOR.DARK,
    },
  },
};

interface ThemePropParams {
  theme: ThemeEnum;
  comp: CompEnum;
  prop: StylePropEnum;
}
const getThemeProp = ({ theme, comp, prop }: ThemePropParams): string => {
  return themeStylePropCollection[theme][comp][prop];
};

const Title = styled.Text<ThemeType>`
  font-size: 12px;
  margin-bottom: 5px;
  color: ${(props): string =>
    getThemeProp({
      theme: props.theme,
      comp: CompEnum.text,
      prop: StylePropEnum.fc,
    })};
`;
const Text = styled.Text<ThemeType>`
  font-size: 14px;
  color: ${(props): string =>
    getThemeProp({
      theme: props.theme,
      comp: CompEnum.text,
      prop: StylePropEnum.fc,
    })};
`;

const RootSelect = styled.View<ThemeType>`
  background-color: ${(props): string =>
    getThemeProp({
      theme: props.theme,
      comp: CompEnum.rootbox,
      prop: StylePropEnum.bc,
    })};
  ${(props): string =>
    getThemeProp({
      theme: props.theme,
      comp: CompEnum.rootbox,
      prop: StylePropEnum.bs,
    })}
  ${(props): string =>
    getThemeProp({
      theme: props.theme,
      comp: CompEnum.rootbox,
      prop: StylePropEnum.border,
    })}
  width: 128px;
  height: 48px;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: 14px 6px;
`;
const ModalBackground = styled.TouchableOpacity`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
`;
const SelectListView = styled.View`
  elevation: 8;
  shadow-color: ${COLOR.DODGERBLUE};
  shadow-offset: {
    width: 0;
    height: 5;
  }
  shadow-opacity: 0.2;
`;
const SelectList = styled(FlatList as new () => FlatList<Item>)`
  background-color: ${COLOR.WHITE};
  padding-top: 8px;
`;

const ItemView = styled.TouchableOpacity<Selected>`
  background-color: ${({ selected }: { selected: boolean }): string =>
    selected ? COLOR.LIGHTBLUE : COLOR.WHITE};
  height: 32px;
  padding: 6px;
  justify-content: center;
`;
const ItemText = styled.Text<Selected>`
  font-size: 14px;
  color: ${COLOR.BLACK};
`;

export type Layout = {
  ox: number;
  oy: number;
  width: number;
  height: number;
};
export interface Props {
  testID?: string;
  items: Item[];
  style?: StyleProp<ViewStyle>;
  theme?: ThemeEnum;
  title?: string;
  selectedValue: string;
  placeholder?: string;
  titleStyle?: StyleProp<TextStyle>;
  activeOpacity: number;
  rootViewStyle?: StyleProp<ViewStyle>;
  rootTextStyle?: StyleProp<TextStyle>;
  mode?: Mode;
  itemListStyle?: ItemListStyle;
  itemViewStyle?: ItemViewStyle;
  itemTextStyle?: StyleProp<TextStyle>;
  showsVerticalScrollIndicator?: boolean;
  disabled?: boolean;
  itemStyle?: StyleProp<ViewStyle>;
  selectedItemStyle?: StyleProp<ViewStyle>;
  onSelect?: (Item) => void;
  onShow?: () => void;
  onDismiss?: () => void;
  onValueChange?: (item: Item, index: number) => void;
}

function Select(props: Props): React.ReactElement {
  const {
    testID,
    theme,
    title,
    titleStyle,
    style,
    activeOpacity,
    disabled,
    mode,
    rootTextStyle,
    rootViewStyle,
    placeholder,
    selectedValue,
    items,
    itemStyle,
    selectedItemStyle,
    onSelect,
  } = props;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const rootEl = useRef<any>();

  const [listOpen, setListOpen] = useState<ListOpen>({
    isOpen: false,
    x: 0,
    y: 0,
    width: 0,
    height: 0,
  });

  const open = useCallback((): void => {
    const nextState = {
      ...listOpen,
      isOpen: true,
    };
    if (rootEl && rootEl.current) {
      rootEl.current.measureInWindow((x, y, width, height) => {
        nextState.x = x;
        nextState.y = y;
        nextState.width = width;
        nextState.height = height;

        setListOpen(nextState);
      });
    }
  }, []);

  const close = useCallback((): void => {
    setListOpen({ ...listOpen, isOpen: false });
  }, []);

  const defaultTheme = disabled
    ? ThemeEnum.disabled
    : !theme
    ? ThemeEnum.none
    : theme;
  const rootViewTheme = disabled
    ? ThemeEnum.disabled
    : rootViewStyle && Object.keys(rootViewStyle).length > 0
    ? ThemeEnum.blank
    : defaultTheme;
  const rootTextTheme = disabled
    ? ThemeEnum.disabled
    : rootTextStyle && Object.keys(rootTextStyle).length > 0
    ? ThemeEnum.blank
    : defaultTheme;
  const titleTextTheme = disabled
    ? ThemeEnum.disabled
    : titleStyle && Object.keys(titleStyle).length > 0
    ? ThemeEnum.blank
    : defaultTheme;
  const _rootViewStyle = disabled ? null : rootViewStyle;
  const _rootTextStyle = disabled ? null : rootTextStyle;

  const renderItem = ({ item }: ListRenderItemInfo<Item>): ReactElement => {
    const style = selectedValue === item.value ? selectedItemStyle : itemStyle;
    return (
      <ItemView
        style={style}
        selected={selectedValue === item.value}
        activeOpacity={1}
        onPress={(): void => {
          handleSelect(item);
        }}
        testID={`${testID}-${TESTID.LISTITEM}-${item.value}`}
      >
        <ItemText selected={selectedValue === item.value} style={style}>
          {item.label}
        </ItemText>
      </ItemView>
    );
  };
  return (
    <View style={style} testID={testID}>
      {title && (
        <Title
          theme={titleTextTheme}
          style={titleStyle}
          testID={`${testID}-${TESTID.TITLETEXT}`}
        >
          {title}
        </Title>
      )}
      <TouchableOpacity
        testID={`${testID}-${TESTID.ROOTBUTTON}`}
        activeOpacity={activeOpacity}
        onPress={open}
        disabled={disabled}
      >
        <RootSelect
          ref={rootEl}
          theme={rootViewTheme}
          style={_rootViewStyle}
          testID={`${testID}-${TESTID.ROOTSELECT}`}
        >
          <Image
            source={!listOpen.isOpen ? IC_ARR_DOWN : IC_ARR_UP}
            testID={`${testID}-${TESTID.ROOTARROW}`}
          />
        </RootSelect>
      </TouchableOpacity>
      <Modal visible={listOpen.isOpen} transparent={true}>
        <ModalBackground onPress={close} />
        <SelectListView
          style={{
            shadowOffset: { width: 0, height: 5 },
            top: listOpen.y,
            left: listOpen.x,
            width: listOpen.width,
            display: listOpen ? 'flex' : 'none',
          }}
          testID={`${testID}-${TESTID.SELECTLISTVIEW}`}
        >
          <SelectList
            style={itemStyle}
            data={items}
            renderItem={renderItem}
            keyExtractor={(item: Item): string => item.value}
          />
        </SelectListView>
      </Modal>
      {/* Todo dropdown */}
      {mode === Mode.dropdown && listOpen.isOpen && (
        <View style={{ height: 50, width: 100, backgroundColor: 'white' }} />
      )}
      {/* Todo dialog */}
      {mode === Mode.dialog && listOpen.isOpen && (
        <View style={{ height: 50, width: 100, backgroundColor: 'white' }} />
      )}
      {mode === Mode.picker && (
        <Picker
          {...props}
          testID={`${testID}-${TESTID.PICKER}`}
          listOpen={listOpen}
          setListOpen={setListOpen}
        />
      )}
    </View>
  );
}

Select.defaultProps = {
  theme: ThemeEnum.none,
  placeholder: '',
  activeOpacity: 0.8,
  mode: Mode.picker,
  showsVerticalScrollIndicator: false,
};

export default Select;
