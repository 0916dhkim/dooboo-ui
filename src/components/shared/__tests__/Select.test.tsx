import 'react-native';

import * as React from 'react';

import { RenderResult, fireEvent, render } from '@testing-library/react-native';
import Select, { Props, TESTID, ThemeEnum } from '../Select';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const ITEMS = [
  { value: 'Category1', text: 'Category1' },
  { value: 'Category2', text: 'Category2' },
  { value: 'Category3', text: 'Category3' },
  { value: 'Category4', text: 'Category4' },
  { value: 'Category5', text: 'Category5' },
];

const mockProp = {
  noTheme: {
    testID: 'select',
    placeholder: 'select',
    rootTextStyle: {
      color: 'orange',
    },
    items: ITEMS,
  },
  inputTheme: {
    testID: 'select',
    theme: ThemeEnum.box,
    placeholder: 'select',
    items: ITEMS,
  },
  themeAndRootView: {
    testID: 'select',
    theme: ThemeEnum.box,
    rootViewStyle: {
      borderBottomColor: 'black',
      borderBottomWidth: 2,
      backgroundColor: 'orange',
    },
    placeholder: 'select',
    items: ITEMS,
  },
  disabled: {
    testID: 'select',
    theme: ThemeEnum.box,
    rootViewStyle: {
      borderBottomColor: 'black',
      borderBottomWidth: 2,
      backgroundColor: 'orange',
    },
    rootTextStyle: {
      color: 'orange',
    },
    placeholder: 'select',
    items: ITEMS,
    disabled: true,
  },
};

interface ObjParam {
  case: number;
  prop: object;
}
const createTestProps = (obj: ObjParam): object => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...mockProp[obj.case],
  ...obj.prop,
});

describe('[Select] render', () => {
  let props: Props;
  let component: React.ReactElement;
  beforeEach(() => {
    props = createTestProps({});
    component = <Select {...props} />;
  });

  it('renders without crashing', () => {
    const rendered: renderer.ReactTestRendererJSON | null = renderer
      .create(component)
      .toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  describe('interactions', () => {
    let testingLib: RenderResult;

    beforeEach(() => {
      testingLib = render(component);
    });

    it('check theme and rootTextStyle and placeholder with case "notheme"', () => {
      const theme = 'noTheme';
      const props = createTestProps({ case: theme });
      const component = <Select {...props} />;
      const testingLib = render(component);
      const selectRoot = testingLib.getByTestId(
        `${props.testID}-${TESTID.ROOTSELECT}`,
      );
      const selectRootText = testingLib.getByTestId(
        `${props.testID}-${TESTID.ROOTTEXT}`,
      );
      const [inputtedRootTextStyle] =
        selectRootText.props.style &&
        selectRootText.props.style.filter((style) => {
          return style === mockProp[theme].rootTextStyle;
        });
      expect(selectRoot.props.theme).toEqual(ThemeEnum.none);
      expect(selectRootText.props.theme).toEqual(ThemeEnum.blank);
      expect(selectRootText.props.children).toEqual(
        mockProp[theme].placeholder,
      );
      expect(inputtedRootTextStyle).toEqual(mockProp[theme].rootTextStyle);
    });

    it('check theme and rootTextStyle and placeholder with case "inputTheme"', () => {
      const theme = 'inputTheme';
      const props = createTestProps({ case: theme });
      const component = <Select {...props} />;
      const testingLib = render(component);
      const selectRoot = testingLib.getByTestId(
        `${props.testID}-${TESTID.ROOTSELECT}`,
      );
      const selectRootText = testingLib.getByTestId(
        `${props.testID}-${TESTID.ROOTTEXT}`,
      );
      const [inputtedRootTextStyle] =
        selectRootText.props.style &&
        selectRootText.props.style.filter((style) => {
          return style === mockProp[theme].rootTextStyle;
        });
      expect(selectRoot.props.theme).toEqual(ThemeEnum.box);
      expect(selectRootText.props.theme).toEqual(ThemeEnum.box);
      expect(selectRootText.props.children).toEqual(
        mockProp[theme].placeholder,
      );
      expect(inputtedRootTextStyle).toEqual(mockProp[theme].rootTextStyle);
    });

    it('check theme and rootViewStyle and placeholder with case "themeAndRootView"', () => {
      const theme = 'themeAndRootView';
      const props = createTestProps({ case: theme });
      const component = <Select {...props} />;
      const testingLib = render(component);
      const selectRoot = testingLib.getByTestId(
        `${props.testID}-${TESTID.ROOTSELECT}`,
      );
      const selectRootText = testingLib.getByTestId(
        `${props.testID}-${TESTID.ROOTTEXT}`,
      );
      const [inputtedRootViewStyle] =
        selectRoot.props.style &&
        selectRoot.props.style.filter((style) => {
          return style === mockProp[theme].rootViewStyle;
        });
      expect(selectRoot.props.theme).toEqual(ThemeEnum.blank);
      expect(selectRootText.props.theme).toEqual(ThemeEnum.box);
      expect(selectRootText.props.children).toEqual(
        mockProp[theme].placeholder,
      );
      expect(inputtedRootViewStyle).toEqual(mockProp[theme].rootViewStyle);
    });

    it('check theme & rootViewStyle & rootTextTheme and placeholder with case "disabled"', () => {
      const theme = 'disabled';
      const props = createTestProps({ case: theme });
      const component = <Select {...props} />;
      const testingLib = render(component);
      const selectRoot = testingLib.getByTestId(
        `${props.testID}-${TESTID.ROOTSELECT}`,
      );
      const selectRootText = testingLib.getByTestId(
        `${props.testID}-${TESTID.ROOTTEXT}`,
      );
      const [inputtedRootViewStyle] =
        selectRoot.props.style &&
        selectRoot.props.style.filter((style) => {
          return style === mockProp[theme].rootViewStyle;
        });
      const [inputtedRootTextStyle] =
        selectRootText.props.style &&
        selectRootText.props.style.filter((style) => {
          return style === mockProp[theme].rootTextStyle;
        });
      expect(selectRoot.props.theme).toEqual(ThemeEnum.disabled);
      expect(selectRootText.props.theme).toEqual(ThemeEnum.disabled);
      expect(selectRootText.props.children).toEqual(
        mockProp[theme].placeholder,
      );
      expect(inputtedRootViewStyle).toBeUndefined();
      expect(inputtedRootTextStyle).toBeUndefined();
    });

    // it('should call [setShowMenu] when focused', () => {
    //   console.log('1-2. called');
    //   const textInput = testingLib.getByTestId('input_chat');
    //   textInput.props.onFocus();
    // });

    // it('should call [setShowMenu] when focused', () => {
    //   console.log('1-4. called');
    //   const touchMenu = testingLib.getByTestId('touch_menu');
    //   fireEvent.press(touchMenu);
    // });
  });
});
