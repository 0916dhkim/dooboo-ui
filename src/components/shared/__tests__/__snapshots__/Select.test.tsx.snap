// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Select] render renders without crashing 1`] = `
<View
  style={
    Array [
      Object {},
    ]
  }
>
  <View
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "elevation": 8,
            "flexDirection": "row",
            "height": 48,
            "justifyContent": "space-between",
            "paddingBottom": 14,
            "paddingLeft": 6,
            "paddingRight": 6,
            "paddingTop": 14,
            "shadowColor": "#3a8bff",
            "shadowOpacity": 0.5,
            "shadowRadius": 5,
            "width": 128,
          },
          null,
        ]
      }
      testID="undefined-root-select"
      theme="none"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#09071d",
              "fontSize": 14,
            },
            null,
          ]
        }
        testID="undefined-root-text"
        theme="none"
      >
        
      </Text>
      <Image
        source={
          Object {
            "testUri": "../../../icons/ic_arr_down.png",
          }
        }
        testID="undefined-root-arrow"
      />
    </View>
  </View>
</View>
`;
